### PR TITLE
[RFR] ra-tree: Components to display hierarchized data

### PR DESCRIPTION
**Pull request by djhi**
Imgur

### CODE REVIEW

**Comment by djhi** [comment]
@fzaninotto Should I add a `Labs` section in the documentation ?

**Comment by djhi** [comment]
> I don't know if implementing a example of a "lab" feature in the simplest example is a good idea.
> New users would be confused IMHO.

The simple example is mainly used for tests, it should probably be renamed. Moreover, I didn't want to introduce yet another example just to test the tree components.

> HS question: what about circular references in the tree ?

Haven't tried that. It's one of those corner cases we'll see when we have time

> This lacks of E2E tests, but .. meh, it's a not beta feature.

Absolutly. It lacks tests, not only e2e

**Comment by Kmaschta** [comment]
I agree. `examples/simple` is definitely not simple.

File: packages/ra-tree-ui-materialui/src/DragPreview.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        } = this.props;
+        return (
+            <div className={className || classes.item} style={style}>
+                {children
+                    ? typeof children === 'function'
+                        ? children({ node, translate })
+                        : children
+                    : translate('ra.tree.drag_preview', {
+                          _:
+                              'Node #%{id} |||| Node #%{id} with %{smart_count} children',
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/DragPreview.js]: well, that's the first time we do such a thing. I don't think we should put the default translation in the component. Makes me wonder how we should incorporate translations from optional packages. Any hint?

File: packages/ra-tree-ui-materialui/src/IgnoreFormProps.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    handleSubmit,
+    handleSubmitWithRedirect,
+    invalid,
+    pristine,
+    saving,
+    submitOnEnter,
+    ...rest
+}) => rest;
+
+const IgnoreFormProps = ({ children, ...props }) => (
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/IgnoreFormProps.js]: I think a jsDoc block would help explaining why and when to use it

File: packages/ra-tree-ui-materialui/src/RootDropTarget.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                    [classes.hover]: canDrop && isOverCurrent,
+                })}
+                disabled={!canDrop}
+            >
+                <IconGetApp />
+                {connectDropTarget(
+                    <div>
+                        <Typography className={classes.text}>
+                            {translate('ra.tree.root_target', {
+                                _: 'Drop an item here to make it top level',
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/RootDropTarget.js]: same remark

File: packages/ra-tree-ui-materialui/src/Tree.js [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        <TreeShowLayout actions={<TreeActions />}>
+            <TextField source="name" />
+        </TreeShowLayout>
+    </Tree>
+
+    // Or
+
+    const TreeActions = props => (
+        <TreeNodeActions {...props}>
+            <SaveButton variant="flat" />
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/Tree.js]: missing decorator

File: packages/ra-tree-ui-materialui/src/Tree.js [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+                                        root: classes.node || undefined,
+                                    }}
+                                    getTreeState={getTreeState}
+                                    node={node}
+                                    treeNodeComponent={TreeNode}
+                                    treeNodeWithChildrenComponent={
+                                        treeNodeWithChildrenComponent
+                                    }
+                                    treeNodeContentComponent={TreeNodeContent}
+                                    {...sanitizeRestProps(props)}
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/Tree.js]: bad, you already cloned the other props in the `TreeController`

File: packages/ra-tree-ui-materialui/src/TreeForm.js [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+                {Children.map(
+                    children,
+                    field =>
+                        field
+                            ? cloneElement(field, {
+                                  basePath: field.props.basePath || basePath,
+                                  onDrop: this.handleDrop,
+                                  record: node.record,
+                                  resource,
+                                  ...sanitizeRestProps(props),
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/TreeForm.js]: I don't think you should pass the additional props to each child, but to the root element

File: packages/ra-tree-ui-materialui/src/TreeNodeWithChildren.js [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+                        ))}
+                    </List>
+                </ExpansionPanelDetails>
+            </ExpansionPanel>
+        );
+    }
+}
+
+const TreeNodeWithChildren = ({ isExpanded, ...props }) => (
+    <TreeContext.Consumer>
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/TreeNodeWithChildren.js]: I'm really not sure you need this context. Can't we pass the props down instead?

File: packages/ra-tree-ui-materialui/src/TreeShowLayout.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            >
+                {Children.map(
+                    children,
+                    field =>
+                        field
+                            ? cloneElement(field, {
+                                  basePath: field.props.basePath || basePath,
+                                  record: node.record,
+                                  resource,
+                                  ...sanitizeRestProps(props),
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/TreeShowLayout.js]: same remark as above, rest props should be passed to the root element, not to the children

File: packages/ra-tree-ui-materialui/src/TreeShowLayout.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+                              })
+                            : null
+                )}
+                {actions &&
+                    cloneElement(actions, {
+                        basePath,
+                        record: node.record,
+                        resource,
+                        ...sanitizeRestProps(props),
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/TreeShowLayout.js]: don't pass the same props to different elements

File: packages/ra-tree-ui-materialui/src/constants.js [hunk 10]
```diff
@@ -0,0 +1 @@
+export const DROP_TARGET_TYPE = 'RA/TREE/NODE';
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/constants.js]: this should be in `ra-tree-core`, no?

File: packages/ra-tree-ui-materialui/src/DragPreview.js (hunk 1 above)
Comment (djhi) [reply packages/ra-tree-ui-materialui/src/DragPreview.js]: Export addon translations packages as we do with the core ones and document how to merge them using lodash?

File: packages/ra-tree-ui-materialui/src/TreeForm.js (hunk 6 above)
Comment (djhi) [reply packages/ra-tree-ui-materialui/src/TreeForm.js]: Some of them must be passed, I'll make it more obvious

File: packages/ra-tree-ui-materialui/src/TreeNodeWithChildren.js (hunk 7 above)
Comment (djhi) [reply packages/ra-tree-ui-materialui/src/TreeNodeWithChildren.js]: Sure

File: packages/ra-tree-ui-materialui/src/constants.js (hunk 10 above)
Comment (djhi) [reply packages/ra-tree-ui-materialui/src/constants.js]: I think not, not all UI implementations might allow drag & drop

File: packages/ra-tree-ui-materialui/src/TreeNodeWithChildren.js (hunk 7 above)
Comment (djhi) [reply packages/ra-tree-ui-materialui/src/TreeNodeWithChildren.js]: It was part of my experimentations regarding performances but it has fixed them

File: packages/ra-tree-core/src/TreeController.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    toggleNode,
+    treeState,
+    ...props
+}) => {
+    const availableData = ids.reduce((acc, id) => [...acc, data[id]], []);
+    const tree = getTreeFromArray(Object.values(availableData), parentSource);
+
+    return (
+        <TreeContext.Provider
+            value={{
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-tree-core/src/TreeController.js]: Performances are indeed better with it. I removed the Context altogether though...

File: packages/ra-tree-ui-materialui/README.md [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+
+### <Tree>
+
+The `Tree` component accepts the following props:
+
+- `enableDragAndDrop`: Enable drag & drop. This will show drag handles on every nodes
+- `allowDropOnRoot`: Setting this prop to `true` will add a root drop zone at the top of the list. Dropping node on it will set their parent to null
+- `dragPreviewComponent`: Customize the preview of the currently dragged node by passing your own component. You can leverage the existing `DragPreview` component and adjusts its content. See [DragPreview](#dragpreview)
+- `getTreeFromArray`: The function used to build the tree from the fetched data. It defaults to one using [performant-array-to-tree](https://github.com/philipstanislaus/performant-array-to-tree)
+- `getTreeState`: A function which must return the tree state root from the redux state in case you mounted it on a different key than `tree`. It will be called with a single `state` argument which is the redux state.
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-tree-ui-materialui/README.md]: It might but it was pretty easy to add (borrowed from redux-form). This might be useful for people integrating react-admin inside an existing app.

File: packages/ra-tree-ui-materialui/README.md [hunk 13]
```diff
@@ -0,0 +1,226 @@
+# ra-tree-ui-materialui
+
+A Tree component with [material-ui](https://github.com/mui-org/material-ui) to use with [react-admin](https://github.com/marmelab/react-admin).
+
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-tree-ui-materialui/README.md]: I added one in the documentation. Should I link it here ?

File: packages/ra-tree-ui-materialui/src/DragPreview.js (hunk 1 above)
Comment (djhi) [reply packages/ra-tree-ui-materialui/src/DragPreview.js]: I added an utility in `ra-core` for merging multiple language packages and added two language packages for `ra-tree`

File: packages/ra-tree-ui-materialui/src/TreeForm.js (hunk 6 above)
Comment (djhi) [reply packages/ra-tree-ui-materialui/src/TreeForm.js]: Actually none of the rest props were needed. I removed this line

File: packages/ra-tree-ui-materialui/src/TreeShowLayout.js (hunk 8 above)
Comment (djhi) [reply packages/ra-tree-ui-materialui/src/TreeShowLayout.js]: Actually none of the rest props were needed. I removed this line

File: packages/ra-tree-ui-materialui/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-tree-ui-materialui/README.md]: yep

File: packages/ra-tree-ui-materialui/src/DragLayer.js [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+        classes: PropTypes.object.isRequired,
+        dragPreviewComponent: PropTypes.oneOfType([
+            PropTypes.element,
+            PropTypes.func,
+        ]).isRequired,
+        itemBeingDragged: PropTypes.object,
+    };
+
+    shouldComponentUpdate(nextProps) {
+        return !isEqual(this.props.offset, nextProps.offset);
```
[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-tree-ui-materialui/src/DragLayer.js]: no, I mean you should check if the `beingDragged` prop changed, too.

File: examples/simple/src/i18nProvider.js [hunk 15]
```diff
@@ -9,6 +9,8 @@ export default locale => {
         return messages[locale]();
     }
 
+    console.log({ englishMessages });
```
Comment (fzaninotto) [inline examples/simple/src/i18nProvider.js]: console.log spotted

File: examples/simple/webpack.config.js [hunk 16]
```diff
@@ -66,6 +66,22 @@ module.exports = {
                 'ra-input-rich-text',
                 'src'
             ),
+            'ra-tree-core': path.join(
```
Comment (fzaninotto) [inline examples/simple/webpack.config.js]: aren't you missing the translation packages?

File: packages/ra-tree-core/README.md [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+- `getTreeState`: A function which must return the tree state root from the redux state in case you mounted it on a different key than `tree`. It will be called with a single `state` argument which is the redux state.
+- `children`: A function which will be called with a single object argument having the following props
+  - `tree`: an array of the root nodes. Each node have the following properties:
+    - `children`: an array of its child nodes
+    - `depth`: a number indicating its depth in the hierarchy
+    - `record`: the node's original data
+  - any additional props received by the `TreeController` component
+- `parentSource`: The field used as the parent identifier for each node. Defaults to `parent_id`
+
+### TreeContext
```
Comment (fzaninotto) [inline packages/ra-tree-core/README.md]: This is no longer part of the packagen you should remove it

File: packages/ra-tree-ui-materialui/src/TreeNode.js [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+        justifyContent: 'space-between',
+        margin: 0,
+        padding: 0,
+    },
+    panelSummaryContent: {
+        alignItems: 'center',
+        margin: 0,
+    },
+    panelSummaryExpanded: {
+        margin: '0 !important',
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/TreeNode.js]: is there an alternative?

File: packages/ra-tree-ui-materialui/src/TreeNode.js [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+                        <TreeNodeContent
+                            key={`TreeNodeContent_${node.id}`}
+                            basePath={basePath}
+                            node={node}
+                            resource={resource}
+                            isLeaf={true}
+                            cancelDropOnChildren={!!itemType}
+                            onDrop={
+                                isOver && canDrop
+                                    ? event => {
```
Comment (fzaninotto) [inline packages/ra-tree-ui-materialui/src/TreeNode.js]: can you extract this function?

File: examples/simple/webpack.config.js (hunk 16 above)
Comment (djhi) [reply examples/simple/webpack.config.js]: Added

**Comment by djhi** [comment]
I have one issue left to fix: a warning about unknown prop
